File: algorithms/cpp/powerOfFour/PowerOfFour.cpp
// Source : https://leetcode.com/problems/power-of-four/
// Author : Hao Chen
// Date   : 2016-05-29

/*************************************************************************************** 
 *
 * Given an integer (signed 32 bits), write a function to check whether it is a power 
 * of 4.
 * 
 * Example:
 * Given num = 16, return true.
 * Given num = 5, return false.
 * 
 * Follow up: Could you solve it without loops/recursion?
 * 
 * Credits:Special thanks to @yukuairoy  for adding this problem and creating all test 
 * cases.
 ***************************************************************************************/


class Solution {
public:
    bool isPowerOfFour(int num) {
        static int mask = 0b01010101010101010101010101010101;
        
        //edge case
        if (num<=0) return false;
        
        // there are multiple bits are 1
        if ((num & num-1) != 0) return false;
        
        // check which one bit is zero, if the place is 1 or 3 or 5 or 7 or 9...,
        // then it is the power of 4
        if ((num & mask) != 0) return true;
        return false;
    }
};

File: algorithms/java/src/PowerOfFour/PowerOfFour.java
// Source : https://leetcode.com/problems/power-of-four/
// Author : Diego Ruiz Piqueras (Pikeras72)
// Date   : 24-04-2022

/***************************************************************************************************** 
 *
 * Given an integer n, return true if it is a power of four. Otherwise, return false.
 * 
 * An integer n is a power of four, if there exists an integer x such that n == 4x.
 *
 * Example 1:
 * 
 * Input: n = 16
 * Output: true
 * 
 * Example 2:
 *
 * Input: n = 5
 * Output: false
 * 
 * Example 3:
 * 
 * Input: n = 1
 * Output: true
 * 
 * Constraints:
 * 
 *  -2^31 <= n <= 2^31 - 1
 * 
 * Explanation of the solution:
 * 
 * After doing the log4(n), if the result have no decimal part, it means that 'n' is a power of four.
 ******************************************************************************************************/

class Solution {
    public boolean isPowerOfFour(int n) {
        double num = Math.log10(n) / Math.log10(4);
        if(num % 1 != 0) {
            return false;
        }
        return true;
    }
}
